lib/services/access_checker.go

```diff
@@ -0,0 +1,460 @@
+	CertificateExtensions() []*types.CertExtension
+
+	// GetSearchAsRoles returns the list of roles which the checker should be able to
+	// "assume" while searching for resources, and should be able to request with a
+	// search-based access request.
+	GetSearchAsRoles() []string
+
+	// MaxConnections returns the maximum number of concurrent ssh connections
+	// allowed.  If MaxConnections is zero then no maximum was defined and the
+	// number of concurrent connections is unconstrained.
+	MaxConnections() int64
+
+	// MaxSessions returns the maximum number of concurrent ssh sessions per
+	// connection. If MaxSessions is zero then no maximum was defined and the
+	// number of sessions is unconstrained.
+	MaxSessions() int64
+
+	// SessionPolicySets returns the list of SessionPolicySets for all roles.
+	SessionPolicySets() []*types.SessionTrackerPolicySet
+
+	// GetAllLogins returns all valid unix logins for the AccessChecker.
+	GetAllLogins() []string
+
+	// GetAllowedResourceIDs returns the list of allowed resources the identity for
+	// the AccessChecker is allowed to access. An empty or nil list indicates that
+	// there are no resource-specific restrictions.
+	GetAllowedResourceIDs() []types.ResourceID
+}
+
+// AccessInfo hold information about an identity necessary to check whether that
+// identity has access to cluster resources. This info can come from a user or
+// host SSH certificate, TLS certificate, or user information stored in the
+// backend.
+type AccessInfo struct {
+	// Roles is the list of cluster local roles for the identity.
+	Roles []string
+	// Traits is the set of traits for the identity.
+	Traits wrappers.Traits
+	// AllowedResourceIDs is the list of resource IDs the identity is allowed to
+	// access. A nil or empty list indicates that no resource-specific
+	// access restrictions should be applied. Used for search-based access
+	// requests.
+	AllowedResourceIDs []types.ResourceID
+	// RoleSet holds the fetched and parsed roles from Roles.
+	RoleSet RoleSet
+}
+
+// AccessCheckerImpl implements the AccessChecker interface.
+type AccessCheckerImpl struct {
+	info         *AccessInfo
+	localCluster string
+
+	// RoleSet is embedded to use the existing implementation for most
+	// AccessChecker methods. Methods which require AllowedResourceIDs (relevant
+	// to search-based access requests) will be implemented by
+	// AccessCheckerImpl.
+	RoleSet
+}
+
+// NewAccessChecker returns a new *AccessCheckerImpl which may be used to check
+// access to resources.
+// Args:
+// - `info *AccessInfo` should at a minimum hold a valid RoleSet for the
+//   identity for which resource access should be checked. It must also hold the
+//   AllowedResourceIDs for the identity if there is any possibility that it has
+//   been granted a search-based access request.
+// - `localCluster string` should be the name of the local cluster in which
+//   access will be checked. You cannot check for access to resources in remote
+//   clusters.
```

Comment:
This makes me wonder how this integrates with the code @Tener for listing allowed db users in `tsh db ls` which [he recently fixed to work with remote clusters](https://github.com/gravitational/teleport/pull/12281).

It mostly uses methods which are defined on `RoleSet` rather than `AccessChecker`.

https://github.com/gravitational/teleport/blob/9377f7c681cf5d17a3cdf1d496ba46638b648fe0/lib/services/role.go#L934

But I think this should be fine as even with search-based access requests, you're temporarily given a new role, so we're going to just read allowed db users from that role?

Reply:
Yeah this should still work, and you can still get at the underlying RoleSet as necessary